File: Python3/DP - 1D/198. House Robber.py
""" 198. House Robber
Difficulty: Medium
Submission Link: https://leetcode.com/problems/house-robber/submissions/1153674627/ """

class Solution:
    def rob(self, nums: List[int]) -> int:
        length=len(nums)
        if length==0:
            return 0
        dp=[0]*length
        if length==1:
            dp[0]=nums[0]
        elif length==2:
            dp[0]=nums[0]
            dp[1]=max(nums[0],nums[1])
        else:
            dp[0] = nums[0]
            dp[1] = max(nums[0], nums[1])
            for i in range(2,length):
                dp[i]=max(dp[i-2]+nums[i],dp[i-1])
        return dp[length-1]

File: C++/DP - 1D/198. House Robber.cpp
/* 198. House Robber
Difficulty: Medium
Submission Link: https://leetcode.com/problems/house-robber/submissions/1153674212/ */

class Solution {
public:
    int rob(vector<int>& nums) {
        if(nums.size() == 0) return 0;
        if(nums.size() == 1) return nums[0];
        if(nums.size() == 2) return max(nums[1],nums[0]);
        if(nums.size() == 3) return max(nums[0]+nums[2], nums[1]);
        int a = nums[0],
        b = max(nums[1],nums[0]),
        c = max(nums[0]+nums[2], nums[1]), 
        temp;
        for(int i=3;i<nums.size();i++){
            temp = max(max(a,b)+nums[i],c);
            a = b;
            b = c;
            c = temp;
        }
        return max(max(a,b),c);
    }
};